src/software/new_geom/rectangle.cpp

```diff
@@ -69,20 +69,20 @@ Point Rectangle::furthestCorner(const Point &p)
         });
 }
 
-bool Rectangle::expand(double amount)
+std::optional<Rectangle> Rectangle::expand(double amount)
 {
     // Ensures rectangle cannot be shrunk to less than a point
     if (xLength() < -2 * amount || yLength() < -2 * amount)
     {
-        return false;
+        return std::nullopt;
     }
 
     points_[0] = points_[0] + Vector(-amount, -amount);
     points_[1] = points_[1] + Vector(-amount, amount);
     points_[2] = points_[2] + Vector(amount, amount);
     points_[3] = points_[3] + Vector(amount, -amount);
 
-    return true;
+    return std::make_optional(*this);
 }
```

Comment:
So this interface is a bit more confusing for me now. Generally I would expect something to have a return value or modify the state of the object, but not both. This is definitely worth having a larger discussion about, as we'll likely see this pattern again for geom stuff.

@jsn0716 @MathewMacDougall @akhilveeraghanta - your thoughts?

Reply:
I agree that both at once is confusing. I don't have a super strong personal preference as long as we're consistent. We already follow the pattern of having a return value and not modifying object state, so I'd probably vote for just sticking with that convention.

Reply:
changed it to return a modified rectangle and I also refactored the failed shrinking case to a bool& because we usually don't care if the expansion succeeded or not (we can just assume it succeeds)

Reply:
I think we may want to stick with returning an optional in all cases though (or making the failure status always available). Right now this assumes we're basically using well-known / constant values as the scaling amount, but this assumption doesn't hold in all cases. Perhaps at some point we dynamically scale based on velocity or some random parameter that could make the scaling positive or negative. In this case we don't know exactly if we will be shrinking or expanding, so would want to check the failure. I think it's easier if we just have one function so developers don't have to be so aware of cases like this.

Reply:
Yeah, I generally agree. I think that:
- If there is an unsafe version that's easier/simpler to use, people will just always use that
- If we're fairly sure the operation is safe, we could just de-reference the returned optional without checking it (which we should really _never_ do)

Reply:
After analysing where scale/expand is being used, I've concluded that most of these situations would benefit from specialized functions. For example:
* Field::getInflatedDefenseArea and Field::getInflatedDefenseArea 
* ObstacleFactory implementing specialized functions for creating expanded rectangles according to robot geometry and constants 
I think that this work is out of scope for integrating obstacles and should be tackled alongside #980

Reply:
I agree that the special cases you highlighted are currently not good uses of `expand`. I'm fine reverting and deferring for now, but can you please make tickets for
* Specifically fixing the `expand` function
* Fixing the special cases you listed

Reply:
#1331

Reply:
Can we make a ticket for re-assessing `expand` so we don't forget?

Reply:
I don't see any point in making a ticket that doesn't have any specific goals or actions. I'm sure we'll remember to do it

Reply:
> I'm sure we'll remember to do it

I don't agree. We've had "investigation" tickets before. It sounds like we might want to do _something_, so I think it's fine for the goal of the ticket to be to define what action we want to take, if any.

Reply:
I've added an extra task for the assignee of #1332 to come up with a new ticket for cleaning up Rectangle::expand, since the assignee would probably know best